File: Tools/1 - Common/MoCommon/System/FSpeedTest.cs
using System;
using System.Collections.Generic;
using System.Text;
using MO.Common.Lang;

namespace MO.Common.System {

   public class FSpeedTest : IDump {

      private TimeSpan _total = new TimeSpan();

      private DateTime _first = DateTime.Now;

      private DateTime _start;

      private DateTime _end;

      private DateTime _last;

      public FSpeedTest() {
         _start = DateTime.Now;
      }

      public TimeSpan Total {
         get { return _total; }
      }

      public DateTime First {
         get { return _first; }
      }

      public DateTime Start {
         get { return _start; }
      }

      public DateTime End {
         get { return _end; }
      }

      public DateTime Last {
         get { return _last; }
      }

      public void DoStart() {
         _start = DateTime.Now;
      }

      public void DoEnd() {
         _end = DateTime.Now;
         _last = DateTime.Now;
         _total += _end.Subtract(_start);
      }

      public TimeSpan GetDiff() {
         return _end.Subtract(_start);
      }

      public string GetDiffSecond() {
         return FormatSpan(GetDiff());
      }

      public string GetTotalSecond() {
         return FormatSpan(_total);
      }

      public string GetTotalNowSecond() {
         return FormatSpan(DateTime.Now.Subtract(_first));
      }

      public string GetNowSecond() {
         return FormatSpan(DateTime.Now.Subtract(_start));
      }

      public string FormatSpan(TimeSpan ts) {
         string totalSeconds = ts.TotalSeconds.ToString();
         int find = totalSeconds.IndexOf('.');
         if (find != -1) {
            string num = totalSeconds.Substring(0, find);
            string aft = totalSeconds.Substring(find + 1);
            if (aft.Length > 3) {
               aft = aft.Substring(0, 3);
            }
            return num + "." + aft.PadRight(3, '0') + "s";
         }
         return totalSeconds + ".000s";
      }

      #region IDump members

      public FDumpInfo Dump() {
         return Dump(new FDumpInfo(this));
      }

      public virtual FDumpInfo Dump(FDumpInfo info) {
         RDump.StartDump(info);
         info.Append("[", _start.ToString("yyMMdd-hhmmss"));
         info.Append(" - ", _end.ToString("yyMMdd-hhmmss"));
         info.Append("] = ", GetDiff().ToString(), "s");
         return info;
      }

      #endregion
   }

}

File: Platform/mo-1-common/src/lang-java/org/mo/com/system/FSpeedTest.java
package org.mo.com.system;

import java.util.Date;

public class FSpeedTest
//implements IDump
{
   public final static String FORMAT_DATE = "YYMMDD-HH24MISS";

   private Date _end;

   private final Date _first = new Date();

   private Date _last;

   private Date _start = new Date();

   private long _total;

   public FSpeedTest(){
   }

   public long diff(){
      return _end.getTime() - _start.getTime();
   }

   public void doEnd(){
      _end = new Date();
      _last = new Date();
      _total += _end.getTime() - _start.getTime();
   }

   public void doStart(){
      _start = new Date();
   }

   //   @Override
   //   public TDumpInfo dump(){
   //      return dump(new TDumpInfo(this));
   //   }
   //   @Override
   //   public TDumpInfo dump(TDumpInfo info){
   //      RDump.append(info);
   //      info.append("[", RDateTime.format(_start, FORMAT_DATE));
   //      info.append(" - ", RDateTime.format(_end, FORMAT_DATE));
   //      //info.append("] = ", GetDiff().ToString(), "s");
   //      return info;
   //   }
   public Date end(){
      return _end;
   }

   public Date first(){
      return _first;
   }

   public String formatDiffSecond(){
      return formatSpan(diff());
   }

   public String formatNowSecond(){
      return formatSpan(new Date().getTime() - _start.getTime());
   }

   public String formatSpan(long span){
      String upper = Long.toString(span / 1000000);
      String lower = Long.toString(span / 1000);
      return upper + "." + lower + "s";
   }

   public String formatTotalNowSecond(){
      return formatSpan(new Date().getTime() - _first.getTime());
   }

   public String formatTotalSecond(){
      return formatSpan(_total);
   }

   public Date last(){
      return _last;
   }

   public Date start(){
      return _start;
   }

   public long total(){
      return _total;
   }
}
